### ISSUE TITLE
Convergence Deltas for DeepLIFT still too high (even after accounting for or deactivating all non-linearities)

### ISSUE THREAD

**Issue by aaronmaiww** [jmschrei/tangermeme#13]
Hey Jacob ,

thanks a lot for this package. I've spent some hours trying to apply DeepLIFT to my CNN but I keep getting convergence deltas that are too high.  

The CNN is pretty simple - it's forward function is just this:

    def forward(self, x):
        x = self.cnn(x)  ### torch.nn.Conv1D
        x = torch.relu(x)

        x = self.post(x)
        return x

Where post(x) is one of three different "heads" (implemented in separate classes), e.g.:

    def forward(self, x):
        sum_p = torch.sum(x, dim=2)
        std_p = torch.std(x, dim=2)
        x = torch.cat((sum_p, std_p), axis=1)
        x = self.bn(x) ## batchnorm
        return x 

I listed the following non-linearities:

additional_nonlinear_ops = {torch.nn.BatchNorm1d: _nonlinear,
                                torch.nn.ReLU: _nonlinear}

And deactivated drop-out with model.eval() -- any thoughts on why I might still get conv deltas > 100?

Thanks so much for any help!!

**Comment by jmschrei** [comment]
Howdy

Sorry you're encountering issues. You can't use `torch.relu` -- every operation that changes values has to be a layer that is registered in the `__init__` function _and only used once._ So, you need one `torch.nn.ReLU()` object per activation in the the network, e.g., the same number of convolution objects and activation objects.

let me know if that fixes your issues.

**Comment by aaronmaiww** [comment]
Thank you for your comment! I tried putting all activations into the __init__ as you say but I still get the same issue. 

Do you think the problem is that my model isn't defined in a single class but rather 2-3 classes (each a component of the model) that are wrapped together by nn.Sequential()? 

The structure is roughly: 

model = CNN(...)
head = nn.Sequential(...) 
joined_model = nn.Sequential(model, head)

**Comment by jmschrei** [comment]
Unfortunately, the DeepLIFT code is very particular when it comes to how models have to be specified so it's hard for me to diagnose from a general description. If you could post the entire code, that would be helpful. It's likely another small issue like the `torch.relu` one.

**Comment by aaronmaiww** [comment]
Appreciate the offer! I tried rewriting the model class such that every layer is defined in the '__init__()'. Here's the outline:

class AllinOne(torch.nn.Module):
    def __init__(self, k, num_filters, sigmoid=False, additional_layer=False, in_channels=4):
        super(AllinOne, self).__init__()
        self.sigmoid = sigmoid
        self.cnn = torch.nn.Conv1d(in_channels=in_channels, out_channels=num_filters, kernel_size=k, bias=biases)
        self.activation1 = torch.nn.Tanh() if sigmoid else torch.nn.ReLU()
        self.bn_2 =nn.BatchNorm1d(...) 

        # MLP 
        self.drop = nn.Dropout(...)
        self.linear = torch.nn.Linear(...)

    def forward(self, x):
        # simple CNN
        x = self.cnn(x)
        x = self.activation1(x)

      
        # new block: head/pooling alternative
        sum_p = torch.sum(x, dim=2)
        std_p = torch.std(x, dim=2)
        x = torch.cat((sum_p, std_p), axis=1)
        x = self.bn_2(x)

        # new pt: dropout + MLP
        x = self.drop(x)
        x = self.linear(x)

        return x

Thanks in advance :)

**Comment by jmschrei** [comment]
Hm. I'm not sure I see anything obvious from your code. I don't think you need to register BatchNorm1d though -- that's just a linear rescaling of your data, right? Also, are you using `self.sigmoid` anywhere? I've had some trouble getting sigmoid working in the past sometimes.

**Comment by aaronmaiww** [comment]
I don't use the sigmoid and, true, batchnormd1d seems linear - just removed it from the non-linear-ops but still get conv deltas >> 100

**Comment by jmschrei** [comment]
Why are you overwriting `torch.nn.ReLU`'s nonlinear function?

**Comment by aaronmaiww** [comment]
```
_nonlinear = lambda x, grad_input, grad_output: grad_input
additional_nonlinear_ops = {torch.nn.ReLU: _nonlinear}
```

You are referring to the above? If yes: I wasn't aware that I was overriding the nn.ReLU function - I seem to have just misunderstood from the tutorial that all non-linearities need to be registered this way. 

I tried removing these two lines and the error persists unfortunately.

**Comment by ruochiz** [comment]
I could be wrong, but looks like `std_p = torch.std(x, dim=2)` this part involves calculating elementwise x^2 during the process, and hence has some nonlinearity in it. 
If you remove that or just replace it with torch.mean, would the delta still be high?

**Comment by aaronmaiww** [comment]
Great catch @ruochiz , thank you so much - this worked!
